docs/static/schemas/draft/2-0-0/application.schema.json

```diff
@@ -0,0 +1,1564 @@
+              "type": "number",
+              "default": 0.9,
+              "examples": [
+                0.5,
+                0.9,
+                0.99
+              ],
+              "minimum": 0,
+              "exclusiveMaximum": 1
+            },
+            "parameters": {
+              "title": "Hardware Parameters",
+              "type": "object",
+              "patternProperties": {
+                "^[a-z][a-z0-9_]*[a-z0-9]$": {
+                  "title": "Hardware Parameter",
+                  "description": "A hardware parameter in the URDF to be overridden with a new value",
+                  "type": "string"
+                }
+              }
+            },
+            "display_name": {
+              "title": "Hardware Display Name",
+              "description": "The human-readable name to display on the hardware interface",
+              "type": "string"
+            },
+            "events": {
+              "title": "Hardware Events",
+              "description": "Hardware event sources that can trigger events",
+              "type": "object",
+              "additionalProperties": false,
+              "properties": {
+                "transitions": {
+                  "title": "Hardware State Transitions",
+                  "description": "Hardware state transitions that can trigger events",
+                  "type": "object",
+                  "additionalProperties": false,
+                  "properties": {
+                    "on_load": {
+                      "$ref": "#/$defs/events"
+                    },
+                    "on_error": {
+                      "$ref": "#/$defs/events"
+                    },
+                    "on_rate_violation": {
+                      "$ref": "#/$defs/events"
+                    },
+                    "on_unload": {
+                      "$ref": "#/$defs/events"
+                    }
+                  }
+                }
+              }
+            },
+            "controllers": {
+              "title": "Controllers",
+              "description": "The controllers attached to the hardware interface",
+              "additionalProperties": false,
+              "patternProperties": {
+                "^[a-z][a-z0-9_]*[a-z0-9]$": {
+                  "title": "Controller",
+                  "description": "A named controller configuration",
+                  "type": "object",
+                  "additionalProperties": false,
+                  "properties": {
+                    "log_level": {
+                      "$ref": "#/$defs/log_level"
+                    },
+                    "plugin": {
+                      "title": "Controller Plugin",
+                      "description": "The name of the controller plugin",
+                      "type": "string"
+                    },
+                    "parameters": {
+                      "$ref": "#/$defs/parameters"
```

Comment:
https://github.com/aica-technology/dynamic-state-engine/issues/278

Do we "allow" the rate parameter for controllers under the `parameters` field like we do for components or do we need to add it as a separate property?

Reply:
good point. If `rate` is a separate field for hardware, and we want that to apply to controllers and components equally, then I think it makes sense to have that as a separate optional field in each case, rather than assuming it under parameters.